mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -1097,6 +1098,440 @@ MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+      toGridBlockSeperation.getStartNames(lowerNameRefs);
+      SmallVector<std::string> upperGridNames;
+      for (unsigned dim = 0; dim < toBeReducedShape.size(); dim++) {
+        upperGridNames.push_back(Twine("grid_dim" + Twine(dim)).str());
+        if (gridSubDims.contains(dim)) {
+          SmallVector<StringRef, 4> upperGridSubDimNames;
+          for (int64_t upperGridSubDim : gridSubDims[dim]) {
+            upperGridSubDimNames.push_back(lowerNameRefs[upperGridSubDim]);
+          }
+          toGridBlockSeperation.merge(upperGridNames.back(), dim,
+                                      upperGridSubDimNames);
+        } else {
+          toGridBlockSeperation.addDim(upperGridNames.back(), dim, 1);
+        }
+      }
+      SmallVector<std::string> upperBlockNames;
+      for (unsigned dim = 0; dim < toBeReducedShape.size(); dim++) {
+        upperBlockNames.push_back(Twine("block_dim" + Twine(dim)).str());
+        if (blockSubDims.contains(dim)) {
+          SmallVector<StringRef, 4> upperBlockSubDimNames;
+          for (int64_t upperBlockSubDim : blockSubDims[dim]) {
+            upperBlockSubDimNames.push_back(lowerNameRefs[upperBlockSubDim]);
+          }
+          toGridBlockSeperation.merge(upperBlockNames.back(),
+                                      dim + toBeReducedShape.size(),
+                                      upperBlockSubDimNames);
+        } else {
+          toGridBlockSeperation.addDim(upperBlockNames.back(),
+                                       dim + toBeReducedShape.size(), 1);
+        }
+      }
+      gridblockSeperation = toGridBlockSeperation.get();
+    }
+    LLVM_DEBUG(llvm::dbgs()
+               << "gridblockSeperation=" << gridblockSeperation << "\n");
+    // Now we join them to finish the recombination.
+    transformAttrs.push_back(gridblockSeperation);
+    transformAttrs.push_back(lastMerge);
+    reduceInShape =
+        cast<TransformMapAttr>(transformAttrs.back()).getLowerBounds();
+    BottomUpTMBuilder dropReductionDim(rewriter, reduceOutShape, loc);
+    for (uint32_t i = 0; i < reduceOutShape.size(); ++i) {
+      if (i == reductionAxis) {
+        dropReductionDim.broadcast({i}, {reduceInShape[i]});
+      } else {
+        dropReductionDim.passThrough({i}, {i});
+      }
+    }
+    transformAttrs.push_back(dropReductionDim.get());
+    threadwiseWriteOp.setExtraViewsAttr(rewriter.getArrayAttr({}));
+    threadwiseWriteOp.getSourceMutable().assign(broadcastReducedSrc);
+    LLVM_DEBUG(llvm::dbgs() << "transformAttrs = "
+                            << "\n";
+               llvm::interleaveComma(transformAttrs, llvm::dbgs());
+               llvm::dbgs() << "\n");
+    TypedValue<ShapedType> reduceOut = reduceOp.getOut();
+    reduceOut = cast<TypedValue<ShapedType>>(
+        applyViewsOnDest(rewriter, loc, reduceOut, transformAttrs));
+    threadwiseWriteOp.getDestMutable().assign(reduceOut);
+    threadwiseWriteOp.setStoreMethodAttr(stMethod);
```

Comment:
is there any case where we could get rid of atomics? maybe leave a TODO as a possible optimization if that's the case.